source/NVDAObjects/UIA/winConsoleUIA.py

```diff
@@ -31,6 +35,121 @@ def __init__(self, obj, position, _rangeObj=None):
 					1
 				)
 
+	def move(self, unit, direction, endPoint=None):
+		oldRange=None
+		if not self._isCaret:
+			# Insure we haven't gone beyond the visible text.
+			# UIA adds thousands of blank lines to the end of the console.
+			visiRanges = self.obj.UIATextPattern.GetVisibleRanges()
+			if visiRanges.length > 0:
+				firstVisiRange = visiRanges.GetElement(0)
+				lastVisiRange = visiRanges.GetElement(visiRanges.length - 1)
+				oldRange=self._rangeObj.clone()
+		if unit == textInfos.UNIT_WORD and direction != 0:
+			# UIA doesn't implement word movement, so we need to do it manually.
+			offset = self._getCurrentOffsetInThisLine()
+			index = 1 if direction > 0 else 0
+			start, end = self._getWordOffsetsInThisLine(offset)
+			wordMoveDirections = (
+				(offset - start) * -1,
+				end - offset
+			)
+			res = self.move(
+				textInfos.UNIT_CHARACTER,
+				wordMoveDirections[index],
+				endPoint=endPoint
+			)
+			if res != 0:
+				return direction
+			else:
+				if self.move(textInfos.UNIT_CHARACTER, -1): # Reset word boundaries to move to the previous word
+					return self.move(unit, direction, endPoint=endPoint)
+				else:
+					return res
+		res = super(consoleUIATextInfo, self).move(unit, direction, endPoint)
+		if oldRange and (
+			self._rangeObj.CompareEndPoints(
+				UIAHandler.TextPatternRangeEndpoint_Start,
+				firstVisiRange,
+				UIAHandler.TextPatternRangeEndpoint_Start
+			) < 0
+			or self._rangeObj.CompareEndPoints(
+				UIAHandler.TextPatternRangeEndpoint_Start,
+				lastVisiRange,
+				UIAHandler.TextPatternRangeEndpoint_End
+			) >= 0
+		):
+			self._rangeObj = oldRange
+			return 0
+		return res
+
+	def expand(self, unit):
+		if unit == textInfos.UNIT_WORD:
+			# UIA doesn't implement word movement, so we need to do it manually.
+			offset = self._getCurrentOffsetInThisLine()
+			start, end = self._getWordOffsetsInThisLine(offset)
+			wordEndPoints = (
```

Comment:
It might make sense to save these as _wordEndPoints on the class, so you don't have to define them twice. Easy to overlook if it ever needs to be changed.

Reply:
`wordEndPoints` and `wordMoveDirections` are slightly different (we subtract one from the former to make `expand` work).

Reply:
Ah, thanks for clarifying.